File: js/gettoniera.js
/**
 * Gettoniera - Simulazione gettoniera per pagamento con monete
 * Gestisce inserimento monete e calcolo importo rimanente
 * 
 * Refactoring: Usa centesimi internamente per evitare errori floating point
 */

class Gettoniera {
  constructor(importoRichiesto = 1.20) {
    // Converti importo richiesto in centesimi
    this.importoRichiestoCents = Math.round(importoRichiesto * 100);
    this.importoInseritoCents = 0;
    this.moneteInserite = []; // Mantiene valori originali in Euro per log/display

    // FEATURE 003: Saldo totale cassetta (non si azzera al reset)
    this.saldoCassettaCents = 0;

    // Tagli monete disponibili (in centesimi)
    this.tagliDisponibiliCents = [100, 50, 20, 10, 5, 2, 1];

    log.info(`💰 Gettoniera inizializzata - Importo richiesto: ${Gettoniera.formattaImporto(this.importoRichiestoCents / 100)}`);
  }

  /**
   * Inserisci una moneta
   * @param {number} valore - Valore moneta in euro (es. 0.50)
   * @returns {boolean} true se inserimento riuscito
   */
  inserisciMoneta(valore) {
    // Converti in centesimi per validazione e calcoli
    const valoreCents = Math.round(valore * 100);

    // Verifica che sia un taglio valido
    if (!this.tagliDisponibiliCents.includes(valoreCents)) {
      log.warn(`⚠️ Taglio moneta non valido: ${Gettoniera.formattaImporto(valore)}`);
      return false;
    }

    // Inserisci moneta (pagamento corrente + cassetta)
    this.importoInseritoCents += valoreCents;
    this.moneteInserite.push(valore);
    this.saldoCassettaCents += valoreCents;

    const rimanente = this.getImportoRimanente();

    log.info(`💰 Moneta inserita: ${Gettoniera.formattaImporto(valore)} | Totale: ${Gettoniera.formattaImporto(this.importoInseritoCents / 100)} | Rimanente: ${Gettoniera.formattaImporto(rimanente)} | Saldo cassetta: ${Gettoniera.formattaImporto(this.saldoCassettaCents / 100)}`);

    // Aggiorna display saldo cassetta
    this.aggiornaSaldoCassetta();

    return true;
  }

  /**
   * Ottieni importo rimanente da versare
   * @returns {number} Importo rimanente in Euro (0 se pagamento completato)
   */
  getImportoRimanente() {
    const rimanenteCents = this.importoRichiestoCents - this.importoInseritoCents;
    return Math.max(0, rimanenteCents / 100);
  }

  /**
   * Verifica se pagamento è completato
   * @returns {boolean} true se importo inserito >= importo richiesto
   */
  isPagamentoCompletato() {
    return this.importoInseritoCents >= this.importoRichiestoCents;
  }

  /**
   * Ottieni importo inserito
   * @returns {number} Importo totale inserito in Euro
   */
  getImportoInserito() {
    return this.importoInseritoCents / 100;
  }

  /**
   * Ottieni lista monete inserite
   * @returns {number[]} Array di valori monete inserite (in Euro)
   */
  getMoneteInserite() {
    return [...this.moneteInserite];
  }

  /**
   * Reset gettoniera (nuovo pagamento)
   * NOTA: Azzera solo importo corrente, NON il saldo cassetta
   */
  reset() {
    const vecchioImportoCents = this.importoInseritoCents;

    this.importoInseritoCents = 0;
    this.moneteInserite = [];
    // NON azzerare this.saldoCassettaCents - rimane in cassetta!

    if (vecchioImportoCents > 0) {
      log.debug(`💰 Gettoniera resettata (importo precedente: ${Gettoniera.formattaImporto(vecchioImportoCents / 100)}) - Saldo cassetta: ${Gettoniera.formattaImporto(this.saldoCassettaCents / 100)}`);
    }
  }

  /**
   * Formatta importo per display
   * @param {number} importo - Importo da formattare in Euro
   * @returns {string} Importo formattato (es. "1,20 €")
   */
  static formattaImporto(importo) {
    return importo.toFixed(2).replace('.', ',') + '€';
  }

  /**
   * Ottieni importo inserito in centesimi (per testing/precisione)
   * @returns {number} Importo in centesimi (intero)
   */
  getImportoInseritoCents() {
    return this.importoInseritoCents;
  }

  /**
   * Ottieni importo rimanente in centesimi (per testing/precisione)
   * @returns {number} Importo rimanente in centesimi (intero)
   */
  getImportoRimanenteCents() {
    return this.importoRichiestoCents - this.importoInseritoCents;
  }

  /**
   * FEATURE 003 (T022): Ottieni saldo monete corrente nella cassetta
   * @returns {number} Saldo totale monete in cassetta (euro)
   */
  getSaldoMonete() {
    const saldo = this.saldoCassettaCents / 100;
    log.debug(`💰 Saldo monete cassetta: ${Gettoniera.formattaImporto(saldo)}`);
    return saldo;
  }

  /**
   * FEATURE 003 (T023): Azzera saldo cassetta monete
   * Simula svuotamento fisico della cassetta
   * @returns {number} Saldo azzerato (per logging operazione)
   */
  azzeraSaldo() {
    const saldoPrecedente = this.getSaldoMonete();

    if (this.saldoCassettaCents <= 0) {
      log.info(`💰 Azzeramento saldo cassetta: già a ${Gettoniera.formattaImporto(0)}`);
      return 0;
    }

    // Azzera SOLO saldo cassetta, non importo corrente
    this.saldoCassettaCents = 0;

    // Aggiorna display saldo cassetta
    this.aggiornaSaldoCassetta();

    log.warn(`💰 Saldo cassetta azzerato: ${Gettoniera.formattaImporto(saldoPrecedente)} → ${Gettoniera.formattaImporto(0)}`);
    return saldoPrecedente;
  }

  /**
   * Aggiorna display saldo cassetta nel pannello admin
   */
  aggiornaSaldoCassetta() {
    const elementoSaldo = document.getElementById('saldo-cassetta-valore');
    if (elementoSaldo) {
      const saldo = this.getSaldoMonete();
      const saldoFormattato = Gettoniera.formattaImporto(saldo);
      elementoSaldo.textContent = `${saldoFormattato}`;
      log.debug(`💰 Display saldo cassetta aggiornato: ${saldoFormattato}`);
    }
  }
}

// Export globale
window.Gettoniera = Gettoniera;

log.info('✅ Gettoniera caricata (versione precisione centesimi)');

File: ios-app/ChioscoApp/Models/Gettoniera.swift
//
//  Gettoniera.swift
//  ChioscoApp
//
//  Modello per simulazione gettoniera
//  Gestisce inserimento monete e calcolo importo rimanente
//  Tradotto da: js/gettoniera.js
//

import Foundation

/// Modello per la gettoniera del chiosco
class GettonierapModello: ObservableObject {
    // Stato pubblicato per SwiftUI
    @Published var importoInserito: Double = 0.0
    @Published var saldoCassetta: Double = 0.0  // Feature 003

    let importoRichiesto: Double
    private var moneteInserite: [Double] = []

    // Tagli monete disponibili (in euro)
    let tagliDisponibili: [Double] = [1.00, 0.50, 0.20, 0.10, 0.05, 0.02, 0.01]

    init(importoRichiesto: Double = 1.20) {
        self.importoRichiesto = importoRichiesto
        Logger.info("💰 Gettoniera inizializzata - Importo richiesto: \(Self.formattaImporto(importoRichiesto))")
    }

    /// Inserisci una moneta
    /// - Parameter valore: Valore moneta in euro
    /// - Returns: true se inserimento riuscito
    func inserisciMoneta(_ valore: Double) -> Bool {
        // Verifica che sia un taglio valido
        guard tagliDisponibili.contains(valore) else {
            Logger.warning("⚠️ Taglio moneta non valido: \(Self.formattaImporto(valore))")
            return false
        }

        // Inserisci moneta (pagamento corrente + cassetta)
        importoInserito += valore
        moneteInserite.append(valore)
        saldoCassetta += valore

        let rimanente = getImportoRimanente()

        Logger.info("💰 Moneta inserita: \(Self.formattaImporto(valore)) | Totale: \(Self.formattaImporto(importoInserito)) | Rimanente: \(Self.formattaImporto(rimanente)) | Saldo cassetta: \(Self.formattaImporto(saldoCassetta))")

        return true
    }

    /// Ottieni importo rimanente da versare
    /// - Returns: Importo rimanente (0 se pagamento completato)
    func getImportoRimanente() -> Double {
        let rimanente = importoRichiesto - importoInserito
        // Arrotonda a 2 decimali per evitare problemi float
        return max(0, (rimanente * 100).rounded() / 100)
    }

    /// Verifica se pagamento è completato
    var isPagamentoCompletato: Bool {
        return getImportoRimanente() <= 0
    }

    /// Ottieni importo inserito
    var importoInseritoCorrected: Double {
        return (importoInserito * 100).rounded() / 100
    }

    /// Ottieni lista monete inserite
    func getMoneteInserite() -> [Double] {
        return moneteInserite
    }

    /// Reset gettoniera (nuovo pagamento)
    /// NOTA: Azzera solo importo corrente, NON il saldo cassetta
    func reset() {
        let vecchioImporto = importoInserito

        importoInserito = 0
        moneteInserite = []
        // NON azzerare saldoCassetta - rimane in cassetta!

        if vecchioImporto > 0 {
            Logger.debug("💰 Gettoniera resettata (importo precedente: \(Self.formattaImporto(vecchioImporto))) - Saldo cassetta: \(Self.formattaImporto(saldoCassetta))")
        }
    }

    /// Ottieni saldo monete corrente nella cassetta (Feature 003)
    /// - Returns: Saldo totale monete in cassetta (euro)
    func getSaldoMonete() -> Double {
        let saldo = saldoCassetta
        Logger.debug("💰 Saldo monete cassetta: \(Self.formattaImporto(saldo))")
        return (saldo * 100).rounded() / 100
    }

    /// Azzera saldo cassetta monete (Feature 003)
    /// Simula svuotamento fisico della cassetta
    /// - Returns: Saldo azzerato (per logging operazione)
    @discardableResult
    func azzeraSaldo() -> Double {
        let saldoPrecedente = getSaldoMonete()

        if saldoPrecedente <= 0 {
            Logger.info("💰 Azzeramento saldo cassetta: già a \(Self.formattaImporto(0))")
            return 0
        }

        // Azzera SOLO saldo cassetta, non importo corrente
        saldoCassetta = 0

        Logger.warning("💰 Saldo cassetta azzerato: \(Self.formattaImporto(saldoPrecedente)) → \(Self.formattaImporto(0))")
        return saldoPrecedente
    }

    /// Formatta importo per display
    /// - Parameter importo: Importo da formattare
    /// - Returns: Importo formattato (es. "1,20€")
    static func formattaImporto(_ importo: Double) -> String {
        return String(format: "%.2f", importo).replacingOccurrences(of: ".", with: ",") + "€"
    }
}
